README.md
[![Build Status](https://travis-ci.org/lirmm/waves-core.svg?branch=master)](https://travis-ci.org/lirmm/waves-core)

README
======

What is WAVES for ?
-------------------

WAVES stands for "\ **W**\ eb \ **A**\ pplication for \ **V**\ ersatile and \ **E**\ asy online bioinformatic \ **S**\ ervices."

WAVES is a dedicated Django based web application to ease bioinformatic tools integration through web interfaces in
order to provide the scientific community with bioinformatic services.

It is aimed to help you easily present, publish and give access on the web to any bioinformatic tool.


Features
--------

- Create and manage your services execution over platform such as Galaxy, DRMAA clusters (SGE), Direct script execution, API calls to other services, remote calls to other platforms via ssh, etc.
- Easily presents these tools in a nice frontend based on Bootstrap3, Bootstrap4 and soon Material Design
- Follow and manage remote REST API access to your service platform
- Manage user's access to your services

.. note::
    - WAVES main component is WAVES-core.
    - WAVES-Galaxy is another component. It is a WAVES adapter to interact with Galaxy instances.
    - WAVES-demo is a custom version of WAVES-core for demo purpose only.

Side projects
-------------

- WAVES-galaxy adapter `<https://www.github.com/lirmm/waves-galaxy>`_
- WAVES-demo project `<https://www.github.com/lirmm/waves-demo>`_


Installation
------------

- You can use WAVES-core as a stand alone application.
- WAVES-core comply to standard reusable project layout for Django. So you may include it as a dependency in your own django project
- Complete documentation available on `readthedocs <https://waves-core.readthedocs.io>`_


Support
-------

If you are having issues, (or just want to say hello): we have a mailing list located at: waves@lirmm.fr


-- UPDATE from <=1.6.4 --
-----------------------------------

If migrating from a version 1.6.4 or earlier to a most recent, you must patch you database encrypted password to a more reliable
encrypted technology (https://github.com/pyca/cryptography). This script must be run once and only once !
To migrate your data please do the following:
- retrieve the latest version
- stop any running service  
- save your database first
- once the new version is installed, run `manage.py update_keys`
- if you don't have any error message: your keys are now more secured ! 


CHANGES.md
CHANGELOG
=========

Version 2.0.0 - 2020-02-25
--------------------------

- [Major] - Some major updates / breaking changes alongside moving to 
    - Python 3 
    - Django 3.0.3
    - Radical saga 1.0.1
- [Major] - Remove swappable for Service/Submission (obsolete use case and work around described [here])
- [Major] - Following packages out of waves.core namespace:
    - api
    - compat
    - ..
- [Major] - deprecated code for WAVES api v1 (will be removed in 3.0)
- [Major] - renamed packages wcore to core


Version 1.6.7 - 2020-01-08
--------------------------

- [Mails] - Ignore settings for admin email - force sending
- [Services Templates] - Integrate JSON-LD into services displayed pages. 


Version 1.6.6 - 2019-09-12
--------------------------

- [Commands] - Fix Error in command to migrate to new encryption lib
- [Queue] - Added Celry/Redis option to process job queue 

Version 1.6.5 - 2019-06-15
--------------------------

- [Encryption] - MAJOR update on cryptography library in use within WAVES - https://github.com/pyca/cryptography see README
- [Commands] - Removed temporarly wqueue processing - issues with remaining processes 


Version 1.6.x - 2018-03-10
----------------------------

- [Layout] - Corrected missing directories
- [ignore] - Added ignored files
- [Bug] - Corrected display in BO for passwords
- [USERS] - Added API USER class, unifying authentication, url redirection dedicated to REST Api users 
    - Added multi-site association allow returning site url prefix for jobs urls
    - Test if job.client user has a 'site' property, in such case, retrieve domain name to generate Job / JobOutputs link
- [LOGGING] JobLogging: degraded mode, output to 'waves.errors' logger
- [Service list] Corrected: list services fails when services are not all public (1.1.9.2).
- [Admin JobList] Corrected: remove wrong filter on jobAdmin queryset (1.1.9.3)
- [API ServiceList] Added: added service_app_name to returned json
- [API Submission] Changed: api_name => submission_app_name in returned json
- [API ServiceList] Changed: changed format for service submission list
 

Version 1.5.x - 2018-02-10
----------------------------

- [JS] Corrected: event association on add new input in submission form
- [Daemon] Corrected: log any fatal exception during job processing in job file
- [Admin Service] Corrected: key error when editing service on 'created_by'
- [Admin Service] Corrected: inline popup on add ExitCode
- [Settings] Added: check for WAVES directories access rights
- [File Permission] Changed to 775/664 job created dirs/files.

Version 1.3 - 2018-02-07
--------------------------

- [Updated] - shared logging behaviour in Jobs / Importer
- [Corrected] - service import method

Version 1.2 - 2018-01-31
--------------------------

- [waves.front] Removed: waves.front, merged back inside wcore
- [Admin] Added: admin urls directly in related Wcore ModelAdmins
- [Models] Added: fields authors and citation to Service model
- [API] Added: Add Job cancel and delete services
- [Models] Removed: removed fields api_on and web_on in Service model
- [Jobs] Changed: JobInconsistentState treatment in job workflow
- [Models] Added: 'topics' and 'operations' property for reading EDAMS ontology data in Service templates
- [Adapters] Moved: Saga common processes in dedicated class SagaAdaptor
- [FileInput] Added: Configuration for enabling/disabling copy/paste form element
- [Docs]: Introduce Administrator guide / Developer guide
- [Api]: Introduce standard token authentication for API form integration
- [Db]: Removed migrations files - causing fails migrate with overridden Submission model - added makemigration upon install

Version 1.1.5 - 2017-11-30
--------------------------

- Corrected: .type property on JobInput
- Added: filter_fields on job ViewSet
- Added: job instance attribute in JobSubmissionViewSet (create_job)
- Added: submission label in Job detail api results

Version 1.1.4 - 2017-10-24
--------------------------

- Corrected fixture loading conflict with signals on api_name duplicate check


Version 1.1.3 - 2017-10-18
--------------------------

- Corrected bugs in Galaxy tool import
- Changed importer API to resolve problem with included runner / adapter params
- added SRV_IMPORT_LOG_LEVEL to configuration

Version 1.1.2 - 2017-09-30
--------------------------

- Added pip package (waves-core) - updated 2017/10/03
- Added changeLog in documentation
- Added BinaryFile upload:
    - Association with Runners / Services / Submission for command lines
- Added standard Django inclusion tags to display Submissions forms templates
    - {% load waves_tags %} => {% submission_form %}, {% service_inc "css" %}, {% service_inc "js" %}
    - Templates directories structure in order to use all available crispy templates packs
    - Check crispy configuration on startup
    - Dynamic includes tags for related css and js from cdnjs
- Simplify overrides for front templates
- Cleaning code to PEP8 standards
- Updated install documentation
- Make Submission model swappable as well
- Added Popup Edit for Submission params Model Admin
- Added quick and dirty solution to override services templates for specific service
- Corrected bugs
    - JsPopupInlines for Django-jet admin layout
    - workflow Runner tests
    - daemon command failed with SQLite DB (depends now on daemons package `̀https://pypi.python.org/pypi/daemons/1.3.0`)


Version 1.1.1 - 2017-07-30
--------------------------

- Corrected many bugs from beta
- api v2
- decoupled front / core templates
- Make Service model swappable (for overriding capabilities in other apps)
- Removed un-mandatory dependencies (django-constance, grappelli, django-jet) - added compat files


Version 1.1.0 - 2017-03-30
--------------------------

Initial Beta version



waves/settings.py
"""
.. See the NOTICE file distributed with this work for additional information
   regarding copyright ownership.
   Licensed under the GNU GPL v3 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at
       https://www.gnu.org/licenses/gpl-3.0.en.html
   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
"""

import logging
import socket
from importlib import import_module
from os.path import join

import six
from django.conf import settings
from django.test.signals import setting_changed

from waves import __version__, __db_version__

try:
    HOSTNAME = socket.gethostname()
except socket.error:
    HOSTNAME = 'localhost'


def perform_import(val, setting_name):
    """
    If the given setting is a string import notation,
    then perform the necessary import or imports.
    """
    if val is None:
        return None
    elif isinstance(val, six.string_types):
        return import_from_string(val)
    elif isinstance(val, (list, tuple)):
        return [import_from_string(item) for item in val]
    return val


def import_from_string(val):
    """
    Attempt to import a class from a string representation.
    """
    try:
        parts = val.split('.')
        module_path, class_name = '.'.join(parts[:-1]), parts[-1]
        module = import_module(module_path)
        return getattr(module, class_name)
    except (ImportError, AttributeError) as e:
        msg = "Could not import '%s' for WAVES settings '%s': %s." % (val, e.__class__.__name__, e)
        raise ImportError(msg)


DEFAULTS = {
    'VERSION': __version__,
    'DB_VERSION': __db_version__,
    'DATA_ROOT': join(getattr(settings, 'BASE_DIR', '/tmp'), 'data'),
    'JOB_BASE_DIR': join(getattr(settings, 'BASE_DIR', '/tmp'), 'data', 'jobs'),
    'BINARIES_DIR': join(getattr(settings, 'BASE_DIR', '/tmp'), 'data', 'bin'),
    'SAMPLE_DIR': join(getattr(settings, 'BASE_DIR', '/tmp'), 'data', 'sample'),
    'UPLOAD_MAX_SIZE': 20 * 1024 * 1024,
    'HOST': HOSTNAME,
    'ADMIN_EMAIL': 'admin@your-site.com',
    'ALLOW_JOB_SUBMISSION': True,
    'APP_NAME': 'WAVES',
    'JOBS_MAX_RETRY': 5,
    'JOB_LOG_LEVEL': logging.INFO,
    'SRV_IMPORT_LOG_LEVEL': logging.INFO,
    'KEEP_ANONYMOUS_JOBS': 30,
    'KEEP_REGISTERED_JOBS': 120,
    'NOTIFY_RESULTS': True,
    'REGISTRATION_ALLOWED': True,
    'SERVICES_EMAIL': 'waves@atgc-montpellier.fr',
    'TEMPLATE_PACK': getattr(settings, 'CRISPY_TEMPLATE_PACK', 'bootstrap3'),
    'SECRET_KEY': getattr(settings, 'SECRET_KEY', '')[0:32],
    'ADAPTORS_CLASSES': (
        'waves.core.adaptors.saga.shell.SshShellAdaptor',
        'waves.core.adaptors.saga.cluster.LocalClusterAdaptor',
        'waves.core.adaptors.saga.shell.SshKeyShellAdaptor',
        'waves.core.adaptors.saga.shell.LocalShellAdaptor',
        'waves.core.adaptors.saga.cluster.SshClusterAdaptor',
        'waves.core.adaptors.saga.cluster.SshKeyClusterAdaptor',
    ),
    'PURGE_WAIT': 86400,
    'PERMISSION_CLASSES': (),
    'MAILER_CLASS': 'waves.core.mails.JobMailer',
}

TEMPLATES_PACKS = ['bootstrap3', 'bootstrap2'],

IMPORT_STRINGS = [
    'ADAPTORS_CLASSES',
    'MAILER_CLASS',
]


def get_db_version():
    return waves_settings.DB_VERSION


class WavesSettings(object):
    """
    WAVES settings object, allow WAVES settings access from properties
    """

    def __init__(self, waves_up_settings=None, defaults=None, imports_string=None):
        if waves_up_settings:
            self._waves_settings = waves_up_settings
        self.defaults = defaults or DEFAULTS
        self.import_strings = imports_string or IMPORT_STRINGS

    @property
    def waves_settings(self):
        if not hasattr(self, '_waves_settings'):
            self._waves_settings = getattr(settings, 'WAVES_CORE', {})
        return self._waves_settings

    def __getattr__(self, attr):
        if attr not in self.defaults:
            raise AttributeError("Invalid WAVES setting: '%s'" % attr)

        try:
            # Check if present in user settings
            val = self.waves_settings[attr]
        except KeyError:
            # Fall back to defaults
            val = self.defaults[attr]

        # Coerce import strings into classes
        if attr in self.import_strings:
            val = perform_import(val, attr)

        # Cache the result
        setattr(self, attr, val)
        return val

    def dump_config(self):
        return sorted([(key, getattr(self, key, self.defaults[key])) for key in self.defaults.keys()])


waves_settings = WavesSettings(None, DEFAULTS, IMPORT_STRINGS)


def reload_waves_settings(*args, **kwargs):
    global waves_settings
    setting, value = kwargs.get('setting'), kwargs.get('value')
    if setting == 'WAVES_CORE':
        waves_settings = WavesSettings(value, DEFAULTS, IMPORT_STRINGS)


setting_changed.connect(reload_waves_settings)
